x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -232,10 +237,33 @@ public void restoreShard(IndexShard indexShard, SnapshotId snapshotId, Version v
         try {
             store.createEmpty(indexShard.indexSettings().getIndexMetaData().getCreationVersion().luceneVersion);
         } catch (EngineException | IOException e) {
-            throw new IndexShardRecoveryException(shardId, "failed to recover from gateway", e);
+            throw new IndexShardRecoveryException(shardId, "failed to create empty store", e);
         } finally {
             store.decRef();
         }
+
+        Store.MetadataSnapshot recoveryMetadata;
+        try {
+            recoveryMetadata = indexShard.snapshotStoreMetadata();
+        } catch (IOException e) {
+            throw new IndexShardRecoveryException(shardId, "failed access store metadata", e);
+        }
+
+        Client remoteClient = client.getRemoteClusterClient(remoteClusterAlias);
+        String sessionUUID = UUIDs.randomBase64UUID();
+        PutCcrRestoreSessionAction.PutCcrRestoreSessionResponse response = remoteClient.execute(PutCcrRestoreSessionAction.INSTANCE,
+            new PutCcrRestoreSessionRequest(sessionUUID, shardId, recoveryMetadata)).actionGet();
```

Comment:
should we have timeouts on these calls (similar as we do for peer recovery within a cluster)? Perhaps something to mark as a follow-up item?

Reply:
I added a todo. I will also add timeout tasks to the meta issue.